### PR TITLE
Refactoring for PLANNER-259

### PR DESCRIPTION
**Pull request by oskopek**

### CODE REVIEW

File: optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/ProblemStatistic.java [hunk 1]
```diff
@@ -120,4 +121,19 @@ protected File writeChartToImageFile(JFreeChart chart, String fileNameBase) {
...
+        if (graphFileList == null || graphFileList.isEmpty()) {
+            return null;
+        } else if (graphFileList.size() > 1) {
+            throw new IllegalStateException("Cannot get graph file for the ProblemStatistic ( " + this
+                    + " ) because it has more than 1 graph file. See method getGraphList() and "
+                    + "ProblemStatisticType.hasScoreLevels()");
+        } else {
+            return graphFileList.get(0);
+        }
+    };
```
Comment (ge0ffrey) [inline optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/ProblemStatistic.java]: the ";" is probably obsolete

File: optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/SingleStatistic.java [hunk 2]
```diff
@@ -150,6 +150,9 @@ public void readCsvStatisticFile() {
             }
             Map<String, String> stringDuplicationRemovalMap = new HashMap<String, String>(1024);
             for (line = reader.readLine(); line != null && !line.isEmpty(); line = reader.readLine()) {
+                if (line.equals("Failed")) { // Reading statistics of a failed benchmark
+                    break;
```
Comment (ge0ffrey) [inline optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/SingleStatistic.java]: I would expect it to do either of these 2 things when it finds "Failed":
- set a flag to mark this single statistic as failed (I presume that it already knows that through the benchmark xml loaded earlier)
- Check if that flag was already set (see above) and fail-fast if it didn't.

File: optaplanner-core/src/main/java/org/optaplanner/core/config/util/ConfigUtils.java [hunk 3]
```diff
@@ -78,6 +78,18 @@
         return ObjectUtils.equals(a, b) ? a : null;
     }
 
+    public static <T> T meldProperty(T a, T b) {
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/config/util/ConfigUtils.java]: The difference between meld and merge isn't inherently clear. Please add a javadoc to explain the difference or rename both methods more self-explanatory.

File: optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/result/ProblemBenchmarkResult.java [hunk 4]
```diff
@@ -453,9 +453,9 @@ public int hashCode() {
                             it.remove();
                         }
                     }
-                    newResult.entityCount = ConfigUtils.mergeProperty(oldResult.entityCount, newResult.entityCount);
-                    newResult.variableCount = ConfigUtils.mergeProperty(oldResult.variableCount, newResult.variableCount);
-                    newResult.problemScale = ConfigUtils.mergeProperty(oldResult.problemScale, newResult.problemScale);
+                    newResult.entityCount = ConfigUtils.meldProperty(oldResult.entityCount, newResult.entityCount);
```
Comment (ge0ffrey) [inline optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/result/ProblemBenchmarkResult.java]: When would a meld (that is not a merge) happen? So when would a.entityCount be null and b.entityCount not be null? I would argue that can only happen when loading older benchmarks, in which the migration code should take care of that (or do nothing and just leave it to become null as merged), and not the agregation code here?

File: optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/ProblemStatistic.java (hunk 1 above)
Comment (oskopek) [inline optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/ProblemStatistic.java]: How did that ever get in there.. Fixed.

File: optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/ProblemStatistic.java [hunk 5]
```diff
@@ -120,4 +121,19 @@ protected File writeChartToImageFile(JFreeChart chart, String fileNameBase) {
...
     }
 
+    public File getGraphFile() {
+        List<File> graphFileList = getGraphFileList();
+        if (graphFileList == null || graphFileList.isEmpty()) {
+            return null;
+        } else if (graphFileList.size() > 1) {
+            throw new IllegalStateException("Cannot get graph file for the ProblemStatistic ( " + this
+                    + " ) because it has more than 1 graph file. See method getGraphList() and "
+                    + "ProblemStatisticType.hasScoreLevels()");
```
Comment (oskopek) [inline optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/ProblemStatistic.java]: Nice trick. Fixed.

File: optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/SingleStatistic.java (hunk 2 above)
Comment (oskopek) [inline optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/SingleStatistic.java]: Fixed with the second approach.

File: optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/result/ProblemBenchmarkResult.java (hunk 4 above)
Comment (oskopek) [inline optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/result/ProblemBenchmarkResult.java]: It was a bug when merging a problem benchmark result with only failed singles. In SingleBenchmarkRunner, `registerScale()` never gets called (for example, the others too). It's version agnostic. So you fix the error, rerun benchmarker, and want to merge those reports. That's when it happens.

Yes, we can leave it to become null, but we know the problem scale from one of them, why not use it? It hasn't changed, the dataset is the same (see `fileToNewResultMap`.. of course, assuming the user didn't change the dataset between runs).

File: optaplanner-core/src/main/java/org/optaplanner/core/config/util/ConfigUtils.java (hunk 3 above)
Comment (oskopek) [inline optaplanner-core/src/main/java/org/optaplanner/core/config/util/ConfigUtils.java]: I couldn't find a more suitable name. Will add the javadoc.

File: optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/result/ProblemBenchmarkResult.java (hunk 4 above)
Comment (ge0ffrey) [inline optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/result/ProblemBenchmarkResult.java]: Makes sense. Would you mind adding a comment to indicate that meld is used to deal with failed singles? I 'll already merge - just add it to another PR.

Comment (oskopek) [inline optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/result/ProblemBenchmarkResult.java]: Fixed in #140.

**Pull request merged**